### PR TITLE
Refactor to better handle expired tokens

### PR DESCRIPTION
**Pull request by jpmartinspt**
## Done
Moved API calls into a separate file, so we can add some extra logic to detect expired macaroons.
Add an handler to clear session and re-login when we detect an expired macaroon.

## QA

- Check out this feature branch
- Run the site using the command `./run serve`
- View the site locally in your web browser at: http://0.0.0.0:8001/
- Make sure `/advantage` still works

Fixes https://github.com/canonical-web-and-design/ubuntu.com/issues/6056

### CODE REVIEW

File: webapp/api/advantage.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    request = Request(method=method, url=f"{API_URL}{path}")
+
+    if session:
+        root = Macaroon.deserialize(flask.session["macaroon_root"])
+        discharge = Macaroon.deserialize(flask.session["macaroon_discharge"])
+        bound = root.prepare_for_request(discharge)
+        token = binary_serialize_macaroons([root, bound]).decode("utf-8")
+        request.headers["Authorization"] = f"Macaroon {token}"
+
+    if data is not None:
```
Comment (anthonydillon) [inline webapp/api/advantage.py]: Can you not do `if data:` here? Why check for None here and not on line 22 above?

File: webapp/api/advantage.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        _prepare_request(
+            method="post",
+            path=f"v1/contracts/{contract_id}/token",
+            data={},
+            session=session,
+        )
+    )
+
+    payload = response.json()
+    return payload.get("contractToken")
```
Comment (anthonydillon) [inline webapp/api/advantage.py]: Why not provide a second param here?

**Comment by jkfran** [comment]
It might be my fault but I got the following error during the QA steps:

```
ERROR gunicorn.error "Exception in worker process"
ModuleNotFoundError: No module named 'werkzeug.middleware'
```
@jpmartinspt ?

File: webapp/api/advantage.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
+import os
+
+import flask
```
Comment (tbille) [inline webapp/api/advantage.py]: not a big fan of putting `flask` dependencies here.
Can you pass the session object that you have to the function that needs it directly.

Comment (jpmartinspt) [reply webapp/api/advantage.py]: Leftover import, already taking the session as a parameter(will remove the places where it's still using flask.session by mistake)

File: webapp/api/advantage.py (hunk 1 above)
Comment (jpmartinspt) [reply webapp/api/advantage.py]: No because we need to call an endpoint with data={} and `{}` would evaluate to `False`.

File: webapp/views.py [hunk 4]
```diff
@@ -60,82 +59,64 @@ def advantage():
...
+                            elif entitlement["type"] == "cc-eal":
+                                entitlements["cc-eal"] = True
+                            elif entitlement["type"] == "support":
+                                contract["supportLevel"] = entitlement[
+                                    "affordances"
+                                ]["supportLevel"]
+                        contract["entitlements"] = entitlements
+                        enterprise_contracts.append(contract)
+        except advantage_api.UnauthorizedRequest:
+            # We got an unauthorized reqest, so we likely
```
Comment (tbille) [inline webapp/views.py]: small typo: `s/reqest/request`

**Comment by anthonydillon** [comment]
QA :+1:

**Pull request merged**